### PR TITLE
Check changelog entries on CI

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
* Improve `check-files.py` to enumerate files with git rather than exclude some ad hoc directory names.
* Tighten `check-files.py` to check all files except ones that are explicitly excluded.
* Fix whitespace and line ending in some files, especially two changelog entries.
* Run `assemble_changelog.py` in `all.sh` so that we know it'll work at release time.

Backports: #3372 (2.16), #3373 (2.7).

### CODE REVIEW

**Comment by mpg** [comment]
>     * Improve `check-files.py` to enumerate files with git rather than exclude some ad hoc directory names.

\o/

File: tests/scripts/check-files.py [hunk 1]
```diff
@@ -290,7 +290,8 @@ def setup_logger(self, log_file, level=logging.INFO):
             console = logging.StreamHandler()
             self.logger.addHandler(console)
 
-    def collect_files(self):
+    @staticmethod
```
Comment (mpg) [inline tests/scripts/check-files.py:301]: The commit message looks off. Did you mean to squash this commit into the previous one?

File: tests/data_files/base64/cli_cid.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
...
+
```
Comment (mpg) [inline tests/data_files/base64/cli_cid.txt:3]: Not a real problem, but it doesn't look like that one was necessary.

File: tests/data_files/base64/cli_no_alpn.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
...
 
+
```
Comment (mpg) [inline tests/data_files/base64/cli_no_alpn.txt:4]: Not a real problem, but it doesn't look like that one was necessary. (8 more instances in the same directory.)

**Comment by mpg** [comment]
Unfortunately the new `all.sh` component failed in the pr-merge job with a `LostContent` exception. I didn't investigate but this needs to be understood and fixed.

**Comment by gilles-peskine-arm** [comment]
A second file with a missing final newline appeared in `development` after I made this PR, that's why the pr-merge job failed. I rebased on top of `development` to be able to fix this file.

File: tests/scripts/check-files.py (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/scripts/check-files.py:301]: Yes.

File: tests/data_files/base64/cli_cid.txt (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/data_files/base64/cli_cid.txt]: I amended the commit to add a trailing newline only when there wasn't one.

**Comment by gilles-peskine-arm** [comment]
Two more files with non-compliant whitespace, another rebase :frowning_face:

**Comment by mpg** [comment]
I've looked at the CI results and the failure in the `ssl-opt.sh` "reconnect from the same port" test is _not_ a resurrection of the failure we used to have in this test: the timeout/resend happens in the initial handshake, so it's quite unrelated to that particular test and could happen in any DTLS test (that asserts no retransmission is expected). The log doesn't really contain enough information to tell, but it could really just be that the server was very slow. I have an item on my list to create a PR that will enhance logging so that we have more information next time this happens.

Anyway, the failure looks entirely unrelated to the present PR and should not block it.

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -625,6 +625,18 @@ component_check_files () {
     record_status tests/scripts/check-files.py
 }
 
+component_check_changelog () {
```
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:628]: This last commit seems quite unrelated to the others. It would probably be better for it to be in a separate PR I think. But as what matters most eventually is the commit history, fine by me if you don't want to do the split.

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:628]: Adding the changelog check was the original goal of this PR. This led into making it pass (fixing the line endings in `ChangeLog.d/*`), then to arranging that it won't fail in an incomprehensible way (modifying `check-files.py` to check the line endings in these files), and from there to fixing the other files where the improved `check-files` found something to say.

File: tests/scripts/check-files.py [hunk 5]
```diff
@@ -26,16 +27,31 @@ class FileIssueTracker:
...
     # pylint: disable=no-member
 
     def __init__(self):
         self.files_with_issues = {}
 
+    @staticmethod
+    def normalize_path(filepath):
+        """Normalize ``filepath`` """
+        filepath = os.path.normpath(filepath)
+        seps = os.path.sep
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/scripts/check-files.py:53]: Because Windows. I'd left the documentation unfinished, I've pushed a new commit to finish it.

**Comment by gilles-peskine-arm** [comment]
> I've looked at the CI results and the failure in the ssl-opt.sh "reconnect from the same port" test

@mpg I observed this on another PR and filed https://github.com/ARMmbed/mbedtls/issues/3377. I haven't looked at the log because some parts of Jenkins are not responding this evening.
